turbopack/crates/turbopack/src/nft_json.rs

```diff
@@ -0,0 +1,194 @@
+    additional_assets: Vec<ResolvedVc<Box<dyn OutputAsset>>>,
+}
+
+#[turbo_tasks::value_impl]
+impl NftJsonAsset {
+    #[turbo_tasks::function]
+    pub fn new(
+        entry: Vc<Box<dyn Module>>,
+        chunk: Option<Vc<Box<dyn OutputAsset>>>,
+        only_externals: bool,
+        output_fs: Vc<DiskFileSystem>,
+        project_fs: Vc<DiskFileSystem>,
+        client_fs: Vc<Box<dyn FileSystem>>,
+        additional_assets: Vec<ResolvedVc<Box<dyn OutputAsset>>>,
+    ) -> Vc<Self> {
+        NftJsonAsset {
+            entry,
+            chunk,
+            only_externals,
+            output_fs,
+            project_fs,
+            client_fs,
+            additional_assets,
+        }
+        .cell()
+    }
+}
+
+#[turbo_tasks::value(transparent)]
+pub struct OutputSpecifier(Option<RcStr>);
+
+#[turbo_tasks::value_impl]
+impl NftJsonAsset {
+    #[turbo_tasks::function]
+    async fn ident_in_project_fs(self: Vc<Self>) -> Result<Vc<FileSystemPath>> {
+        let this = self.await?;
+        let project_fs = this.project_fs.await?;
+        let output_fs = this.output_fs.await?;
+        let nft_folder = self.ident().path().parent().await?;
+
+        if let Some(subdir) = output_fs.root.strip_prefix(&*project_fs.root) {
+            Ok(this
+                .project_fs
+                .root()
+                .join(subdir.into())
+                .join(nft_folder.path.clone()))
+        } else {
+            bail!("TODO output fs not inside project fs");
+        }
+    }
+
+    #[turbo_tasks::function]
+    async fn ident_in_client_fs(self: Vc<Self>) -> Result<Vc<FileSystemPath>> {
+        Ok(self
+            .await?
+            .client_fs
+            .root()
+            .join(self.ident().path().parent().await?.path.clone()))
+    }
+
+    #[turbo_tasks::function]
+    async fn get_output_specifier(
+        self: Vc<Self>,
+        path: Vc<FileSystemPath>,
+    ) -> Result<Vc<OutputSpecifier>> {
+        let this = self.await?;
+        let path_fs = path.fs().resolve().await?;
+        let path_ref = path.await?;
+        let nft_folder = self.ident().path().parent().await?;
+
+        if path_fs == Vc::upcast(this.output_fs.resolve().await?) {
+            // e.g. a referenced chunk
+            return Ok(Vc::cell(Some(
+                nft_folder.get_relative_path_to(&path_ref).unwrap(),
+            )));
+        } else if path_fs == Vc::upcast(this.project_fs.resolve().await?) {
+            return Ok(Vc::cell(Some(
+                self.ident_in_project_fs()
+                    .await?
+                    .get_relative_path_to(&path_ref)
+                    .unwrap(),
+            )));
+        } else if path_fs == Vc::upcast(this.client_fs.resolve().await?) {
+            return Ok(Vc::cell(Some(
+                self.ident_in_client_fs()
+                    .await?
+                    .get_relative_path_to(&path_ref)
+                    .unwrap()
+                    .replace("/_next/", "/.next/")
```

Comment:
We probably should figure out a more general way of handling these, rather than special casing for nextjs